File: src/formats/glyph.ts
// Native .glyph format — lossless JSON serialization of SubtitleDocument.
//
// This is the canonical save format. Everything in the model (cues, tokens,
// styles, meta) round-trips exactly. External formats are lossy "exports".

import {
  GLYPH_SCHEMA_VERSION,
  type GlyphFile,
  type SubtitleDocument,
  emptyDocument,
} from "../types/subtitle";

export function serializeGlyph(doc: SubtitleDocument): string {
  const file: GlyphFile = {
    schemaVersion: GLYPH_SCHEMA_VERSION,
    document: doc,
  };
  return JSON.stringify(file, null, 2);
}

export function parseGlyph(raw: string): SubtitleDocument {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (e) {
    throw new Error(`.glyph 파싱 실패: ${(e as Error).message}`);
  }

  const file = parsed as Partial<GlyphFile>;
  if (!file || typeof file !== "object" || !file.document) {
    throw new Error(".glyph 형식이 올바르지 않습니다 (document 누락).");
  }

  return migrate(file.schemaVersion ?? 1, file.document);
}

/**
 * Apply forward migrations. Unknown fields on the document are preserved as-is
 * (we never strip data we don't recognize). Add cases here as the schema grows.
 */
function migrate(version: number, doc: SubtitleDocument): SubtitleDocument {
  // v1 is the current schema — no transforms yet.
  if (version > GLYPH_SCHEMA_VERSION) {
    // File written by a newer app version. Load defensively rather than fail.
    console.warn(
      `.glyph schemaVersion ${version} is newer than supported ${GLYPH_SCHEMA_VERSION}; loading as-is.`,
    );
  }
  return {
    ...emptyDocument(doc.format ?? "srt"),
    ...doc,
    cues: Array.isArray(doc.cues) ? doc.cues : [],
    meta: doc.meta ?? {},
  };
}

File: native/Sources/GlyphlineCore/Formats/Glyph.swift
// Native .glyph format — lossless JSON serialization of SubtitleDocument
// (ported from ../../src/formats/glyph.ts). This is the canonical save format;
// everything in the model round-trips exactly. External formats are lossy exports.

import Foundation

public enum GlyphError: Error, CustomStringConvertible {
    case invalid(String)
    public var description: String {
        switch self { case .invalid(let m): return m }
    }
}

public func serializeGlyph(_ doc: SubtitleDocument) throws -> String {
    let enc = JSONEncoder()
    enc.outputFormatting = [.prettyPrinted]
    let data = try enc.encode(GlyphFile(document: doc))
    return String(data: data, encoding: .utf8) ?? ""
}

public func parseGlyph(_ raw: String) throws -> SubtitleDocument {
    guard let data = raw.data(using: .utf8) else { throw GlyphError.invalid(".glyph 인코딩 오류") }
    // Lenient wrapper: schemaVersion may be absent (defaults to 1), matching TS.
    struct LenientFile: Decodable {
        var schemaVersion: Int?
        var document: SubtitleDocument
    }
    let file: LenientFile
    do {
        file = try JSONDecoder().decode(LenientFile.self, from: data)
    } catch {
        throw GlyphError.invalid(".glyph 파싱 실패: \(error)")
    }
    return migrate(file.schemaVersion ?? 1, file.document)
}

/// Forward migrations. v1 is current; a newer file loads as-is (defensive).
/// NOTE: unlike the TS version, Swift's decoder drops unrecognized JSON keys —
/// revisit if forward-compat preservation of unknown fields becomes needed.
private func migrate(_ version: Int, _ doc: SubtitleDocument) -> SubtitleDocument {
    _ = version
    return doc
}
